src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -329,10 +318,22 @@ std::optional<edm4eic::MutableCluster> CalorimeterClusterRecoCoG::reconstruct(co
   cl.addToShapeParameters( eigenValues_3D[0].real() ); // 3D x-y-z cluster width 1
   cl.addToShapeParameters( eigenValues_3D[1].real() ); // 3D x-y-z cluster width 2
   cl.addToShapeParameters( eigenValues_3D[2].real() ); // 3D x-y-z cluster width 3
-  //last 3 shape parameters are the components of the axis direction
-  cl.addToShapeParameters( axis_x );
-  cl.addToShapeParameters( axis_y );
-  cl.addToShapeParameters( axis_z );
+
+
+  double dot_product = cl.getPosition() * axis;
+  if (dot_product < 0) {
+    axis = -1 * axis;
+  }
+
+  if (m_cfg.longitudinalShowerInfoAvailable) {
```

Comment:
An easy check was to see if `axis_z != 0`, but that is questionable.

Reply:
This situation is the singular matrix exception, right? Two eigenvalues non-zero and last eigenvalue zero.

Reply:
No it's a possibility for "flat" arrangement, there is no inversion going on.

Reply:
But still, flat arrangement results in a 2D cluster with two principal directions in the detector plane, and the third eigenvector normal to the plane with an eigenvalue of zero (size of cluster in that direction). Having one zero eigenvalue makes this a singular matrix.

Reply:
Yes, singular matrices have eigenvalues. If lowest eigenvalue is zero, that means that cluster centers are on a single plane. This is practically only possible in a "flat" detector that doesn't provide 3d shower profile.

Reply:
As discussed and agreed with you, the preference is for making detectors explicitly 'flat' to avoid having to introduce arbitrary tolerance parameters to allow for 3d clusters solely introduced by minor misalignment in the block perpendicular to their arrangement plane.